### PR TITLE
feat: add event specifications

### PR DESCRIPTION
**Pull request by turip**
## Overview

This patch adds the ingestion and entitlement-based event signatures.

For usage please see the testcase for the parser. Right now we are only validating the balance
snapshot events, but it's possible to do it for other events if needed.

### CODE REVIEW

File: internal/event/types/event_type.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+	// SubjectKind specifies the kind of the subject. Used to construct the subject field in /namespace/<id>/kind/<entryID>
+	SubjectKind string
+
+	// cloudEventType is the actual cloud event type, so that we don't have the calculate it
+	// for each message
+	cloudEventType string
+}
+
+func (s *EventTypeSpec) FillEvent(ev event.Event) {
```
Comment (turip) [inline internal/event/types/event_type.go:31]: Move to NewCloudEvent

File: internal/event/types/event_type.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+	// Time specifies when the event occurred
+	Time time.Time
+
+	// Subject meta (optional, references the entity/subject of the event)
+	Namespace string
+	SubjectID string
+}
+
+func (c *EventSpec) FillEvent(ev event.Event, subjectKind string) {
+	if c.Time.IsZero() {
```
Comment (turip) [inline internal/event/types/event_type.go:60]: Move to NewCloudEvent

**Review by chrisgacsal — CHANGES_REQUESTED** [review]

File: internal/event/types/parser.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+// NewCloudEvent creates a new CloudEvent with the given event spec and payload
+// example usage:
+//
+//	 ev, err := CreateCloudEvent(EventSpec{
+//					ID:    "123",
+//					Source: "test",
+//	 }, IngestEvent{...})
+func NewCloudEvent(eventSpec EventSpec, payload CloudEventsPayload) (event.Event, error) {
+	// Mandatory cloud events fields
```
Comment (turip) [inline internal/event/types/parser.go:26]: Expected source/subject fields:

```
grant:
    source: //openmeter.io/namespace/<>/entitlement/<>/grant/<id>
    subject: //openmeter.io/namespace/<>/subject/<subjectID>

entitlement:
    source: //openmeter.io/namespace/<>/entitlement/<>
    subject: //openmeter.io/namespace/<>/subject/<subjectID>

ingest:
    source: //openmeter.io/namespace/<>/event
    subject: //openmeter.io/namespace/<>/subject/<subjectID>
```

File: internal/event/schema/entitlement_balance_snapshot.go [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+	// Overage Only available for metered entitlements. Overage represents the usage that wasn't covered by grants, e.g. if the subject had a total feature usage of 100 in the period but they were only granted 80, there would be 20 overage.
+	Overage *float64 `json:"overage,omitempty"`
+
+	// Usage Only available for metered entitlements. Returns the total feature usage in the current period.
+	Usage *float64 `json:"usage,omitempty"`
+}
+
+type EntitlementBalanceSnapshotEvent struct {
+	EntitlementID string          `json:"entitlementId"`
```
[the remark this answers is not in the document]
Comment (turip) [reply internal/event/schema/entitlement_balance_snapshot.go]: I haven't expanded the subject, as it's a bit tricky in terms of cloud vs openmeter, others are done.

File: internal/credit/events.go [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+	if err := g.Namespace.Validate(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+type GrantCreatedEvent grantEvent
+
+var grantCreatedEventSpec = spec.EventTypeSpec{
```
Comment (chrisgacsal) [inline internal/credit/events.go:50]:
```suggestion
var grantCreatedEventSpec = spec.EventTypeSpec {
```

File: internal/credit/events.go [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+	return &grantCreatedEventSpec
+}
+
+func (e GrantCreatedEvent) Validate() error {
+	return grantEvent(e).Validate()
+}
+
+type GrantVoidedEvent grantEvent
+
+var grantVoidedEventSpec = spec.EventTypeSpec{
```
Comment (chrisgacsal) [inline internal/credit/events.go:67]:
```suggestion
var grantVoidedEventSpec = spec.EventTypeSpec {
```

File: internal/entitlement/events.go [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+	if err := e.Namespace.Validate(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+type EntitlementCreatedEvent entitlementEvent
+
+var entitlementCreatedEventSpec = spec.EventTypeSpec{
```
Comment (chrisgacsal) [inline internal/entitlement/events.go:42]:
```suggestion
var entitlementCreatedEventSpec = spec.EventTypeSpec {
```

File: internal/entitlement/events.go [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+	return &entitlementCreatedEventSpec
+}
+
+func (e EntitlementCreatedEvent) Validate() error {
+	return entitlementEvent(e).Validate()
+}
+
+type EntitlementDeletedEvent entitlementEvent
+
+var entitlementDeletedEventSpec = spec.EventTypeSpec{
```
Comment (chrisgacsal) [inline internal/entitlement/events.go:59]:
```suggestion
var entitlementDeletedEventSpec = spec.EventTypeSpec {
```

File: internal/entitlement/metered/events.go [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+
+type ResetEntitlementEvent struct {
+	EntitlementID string                 `json:"entitlementId"`
+	Namespace     models.NamespaceID     `json:"namespace"`
+	Subject       models.SubjectKeyAndID `json:"subject"`
+	ResetAt       time.Time              `json:"resetAt"`
+	RetainAnchor  bool                   `json:"retainAnchor"`
+}
+
+var resetEntitlementEventSpec = spec.EventTypeSpec{
```
Comment (chrisgacsal) [inline internal/entitlement/metered/events.go:28]:
```suggestion
var resetEntitlementEventSpec = spec.EventTypeSpec {
```

File: internal/entitlement/snapshot/event.go [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+
+	// CalculatedAt specifies when the balance calculation was performed. It can be used to verify
+	// in edge-worker if the store already contains the required item.
+	CalculatedAt *time.Time `json:"calculatedAt,omitempty"`
+
+	Balance            *EntitlementValue  `json:"balance,omitempty"`
+	CurrentUsagePeriod *recurrence.Period `json:"currentUsagePeriod,omitempty"`
+}
+
+var entitlementBalanceSnapshotEventSpec = spec.EventTypeSpec{
```
Comment (chrisgacsal) [inline internal/entitlement/snapshot/event.go:58]:
```suggestion
var entitlementBalanceSnapshotEventSpec = spec.EventTypeSpec {
```

File: internal/sink/events.go [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+	Namespace  models.NamespaceID `json:"namespace"`
+	SubjectKey string             `json:"subjectKey"`
+
+	// MeterSlugs contain the list of slugs that are affected by the event. We
+	// should not use meterIDs as they are not something present in the open source
+	// version, thus any code that is in opensource should not rely on them.
+	MeterSlugs []string `json:"meterSlugs"`
+}
+
+var ingestEventSpec = spec.EventTypeSpec{
```
Comment (chrisgacsal) [inline internal/sink/events.go:28]:
```suggestion
var ingestEventSpec = spec.EventTypeSpec {
```

File: internal/entitlement/metered/events.go [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+	return nil
+}
+
+type CreateMeteredEntitlementGrantEvent struct {
+	EntitlementGrant
+	Subject   models.SubjectKeyAndID `json:"subjectKey"`
+	Namespace models.NamespaceID     `json:"namespace"`
+}
+
+var createMeteredEntitlementGrantEventSpec = spec.EventTypeSpec{
```
Comment (chrisgacsal) [inline internal/entitlement/metered/events.go:64]:
```suggestion
var createMeteredEntitlementGrantEventSpec = spec.EventTypeSpec {
```

**Review by chrisgacsal — APPROVED** [review]

**Pull request merged**